src/realizations/catchment/Bmi_C_Formulation.cpp

```diff
@@ -0,0 +1,161 @@
+    return "bmi_c";
+}
+
+void Bmi_C_Formulation::create_formulation(boost::property_tree::ptree &config, geojson::PropertyMap *global) {
+    geojson::PropertyMap options = this->interpret_parameters(config, global);
+    inner_create_formulation(options, false);
+}
+
+void Bmi_C_Formulation::create_formulation(geojson::PropertyMap properties) {
+    inner_create_formulation(properties, true);
+}
+
+std::string Bmi_C_Formulation::get_output_header_line(std::string delimiter) {
+    return boost::algorithm::join(get_output_header_fields(), delimiter);
+}
+
+std::string Bmi_C_Formulation::get_output_line_for_timestep(int timestep, std::string delimiter) {
+    // Check if the timestep is in bounds for the fluxes vector, and handle case when it isn't
+    if (timestep > get_bmi_model()->get_last_processed_time_step()) {
+        return "";
+    }
+    std::string output_str;
+
+    for (const std::string& name : get_output_variable_names()) {
+        output_str += (output_str.empty() ? "" : ",") + std::to_string(get_var_value_as_double(timestep, name));
+    }
+    return output_str;
+}
+
+/**
+ * Get the model response for this time step.
+ *
+ * Get the model response for this time step, execute the backing model formulation one or more times if the time step
+ * of the given index has not already been processed.
+ *
+ * Function assumes the backing model has been fully initialized an that any additional input values have been applied.
+ *
+ * The function will return the value of the primary output variable (see `get_bmi_main_output_var()`) for the given
+ * time step. The type returned will always be a `double`, with other numeric types being cast if necessary.
+ *
+ * Because of the nature of BMI, the `t_delta` parameter is ignored, as this cannot be passed meaningfully via
+ * the `update()` BMI function.
+ *
+ * @param t_index The index of the time step for which to run model calculations.
+ * @param d_delta_s The duration, in seconds, of the time step for which to run model calculations.
+ * @return The total discharge of the model for the given time step.
+ */
+double Bmi_C_Formulation::get_response(time_step_t t_index, time_step_t t_delta) {
+    if (get_bmi_model() == nullptr) {
+        throw std::runtime_error("Trying to process response of improperly created BMI C formulation.");
+    }
+    if (t_index < 0) {
+        throw std::invalid_argument("Getting response of negative time step in BMI C formulation is not allowed.");
+    }
+
+    int last_processed_time_step = get_bmi_model()->get_last_processed_time_step();
+    while (last_processed_time_step++ < t_index) {
+        get_bmi_model()->Update();
+    }
+    return get_var_value_as_double(t_index, get_bmi_main_output_var());
+}
+
+double Bmi_C_Formulation::get_var_value_as_double(time_step_t t_index, const std::string& var_name) {
+    // TODO: change to using "at index" option
+    std::string type = get_bmi_model()->GetVarType(var_name);
+    if (type == "double") {
```

Comment:
Should probably make an enum for "types" and do this switch on the enum int values instead of string comparisons (or a `bitstring`, cause `bitstring`'s!).

Reply:
There's really not a way around doing the string comparison, since BMI specifies that the `get_var_type` function specification returns a string.  There are other ways we could organize it (e.g., a map of strings to enum values), but I'm not sure any really serve as an improvement.

The exception would be if we are also storing this for later lookups also, though that may or may not be a good direction to go in.  Although, I'm liking this method less and less the more I look at it.

I did go ahead and add more of the standard types to the logic, so it's more gnarly looking at least.